orttraining/orttraining/training_api/module.cc

```diff
@@ -48,62 +49,103 @@ Status Parameter::ResetGrad() {
-    eval_sess_ = std::make_unique<onnxruntime::InferenceSession>(so, *env);
-    ORT_THROW_IF_ERROR(eval_sess_->Load(eval_model_path_or_bytes.value()));
-    ORT_THROW_IF_ERROR(eval_sess_->Initialize());
-    utils::GetGraphInputOutputNames(eval_sess_, eval_input_names_, eval_output_names_);
-  }
 
+Module::Module(std::unordered_map<std::string, std::shared_ptr<Parameter>>& named_parameters,
+               InferenceSession* train_session) {
+  train_sess_ = train_session;
   utils::GetGraphInputOutputNames(train_sess_, train_input_names_, train_output_names_);
+
   auto& train_sess_state = train_sess_->GetSessionState();
   std::vector<std::string> param_input_names, grad_input_names, user_input_names, reset_grad_name;
   std::string param_name;
 
+  std::unordered_map<std::string, size_t> param_name_to_grad_input_index_map;
   for (const auto& input_name : train_input_names_) {
-    auto it = named_parameters_.find(input_name);
-    if (it != named_parameters_.end()) {
+    auto it = named_parameters.find(input_name);
+    if (it != named_parameters.end()) {
       param_input_names.emplace_back(input_name);
-      weights_.push_back(it->second->Data());
     } else if (input_name == ACCUMULATE_GRAD_CONTROL_INPUT_NAME) {
       reset_grad_name.emplace_back(input_name);
     } else if (utils::GetParamNameFromGradient(input_name, param_name)) {
+      param_name_to_grad_input_index_map.insert({param_name, grad_input_names.size()});
       grad_input_names.emplace_back(input_name);
-      // create gradient buffer
-      auto it = named_parameters_.find(param_name);
-      ORT_ENFORCE(it != named_parameters_.end(), "Unknown param: ", param_name, " for gradient input: ", input_name);
-      // TODO: don't pre-allocate the gradient buffer.
-      ORT_THROW_IF_ERROR(it->second->AllocateGrad(input_name, train_sess_state));
-      gradients_.push_back(it->second->Gradient());
     } else {
       user_input_names.emplace_back(input_name);
     }
   }
+
+  gradients_.resize(grad_input_names.size());
+
   train_input_names_ = user_input_names;
   train_input_names_.insert(train_input_names_.end(), param_input_names.begin(), param_input_names.end());
   train_input_names_.insert(train_input_names_.end(), grad_input_names.begin(), grad_input_names.end());
   train_input_names_.insert(train_input_names_.end(), reset_grad_name.begin(), reset_grad_name.end());
 
-  // Eval model validation
-  if (nullptr != eval_sess_) {
+  // Loop each parameter, allocate it's memory based on user specified device.
+  for (auto& param_name : param_input_names) {
+    ORT_ENFORCE(named_parameters.find(param_name) != named_parameters.end());
+    OrtValue& source_ortvalue = named_parameters[param_name]->Data();
+    ORT_ENFORCE(source_ortvalue.IsTensor());
+    const Tensor& source_tensor = source_ortvalue.Get<Tensor>();
+
+    std::vector<SessionState::NodeInfo> node_info_vec;
+    ORT_THROW_IF_ERROR(train_sess_state.GetInputNodeInfo(param_name, node_info_vec));
```

Comment:
since this is called after session initialize, the device partitioning is already done. Curious if the memcpy nodes are added for input edges already- and if yes, this NodeInfo is for the memcpy node or the consumer node?

Reply:
This is a good question! 
In observation there is no memcpy node. so I think it comes from consumer node. Will keep monitoring in profiling more models.